components/script/dom/bindings/str.rs

```diff
@@ -208,6 +208,95 @@ impl DOMString {
         self.0.truncate(last_non_whitespace);
         let _ = self.0.splice(0..first_non_whitespace, "");
     }
+
+    /// Valid date string should be YYYY-MM-DD, YYYY must be four or more
+    /// https://html.spec.whatwg.org/multipage/#valid-date-string
+    pub fn is_valid_date_string(&mut self) -> bool {
+        // Step 2
+        let value = &*(self.0);
+        let value_vec: Vec<&str> = value.split('-').collect();
+        if !(value_vec.len() == 3) {
+            return false;
+        }
+        let year = value_vec[0];
+        let month = value_vec[1];
+        let day = value_vec[2];
+
+        // Step 1
+        let month_string = format!("{}-{}", year, month);
+        if !self.check_valid_month_string(&*month_string) {
+            return false;
+        }
+
+        // Step 3
+        if !(day.len() == 2 && day.chars().all(|c| c.is_digit(10))) {
+            return false;
+        }
+        let year_int = year.parse::<i32>().unwrap();
+        let month_int = month.parse::<i32>().unwrap();
+        let day_int = day.parse::<i32>().unwrap();
```

Comment:
Sorry if I wasn't clear on my last request, but what I mean is that `check_valid_month_string` should instead be `parse_month_string`, which returns a type of `Result<(u32, u32), ()>` that contains a 2-tuple of a year-month pair for the success variant, and just an empty error for the other variant. That way, you don't need to parse here.

Reply:
@KiChjang 
There is still `year` need to parse:
```
let year_int = year.parse::<i32>().unwrap();
```
If we parse the three(year, month, day) together would let the code more clear?

Reply:
That's a code duplication that I would like to avoid, since you should have already parsed the month and year in the `parse_month_string` algorithm.